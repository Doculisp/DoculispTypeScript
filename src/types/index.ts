// Export all types for external consumption
export * from './types.ast';
export * from './types.astDoculisp';
export * from './types.astProject';
export * from './types.containers';
export * from './types.controller';
export * from './types.document';
export * from './types.docWriter';
export * from './types.fileHandler';
export * from './types.filePath';
export * from './types.general';
export * from './types.includeBuilder';
export * from './types.internal';
export * from './types.sringBuilder';
export * from './types.stringWriter';
export * from './types.structure';
export * from './types.textHelpers';
export * from './types.tokens';
export * from './types.trimArray';
export * from './types.variableTable';
export * from './types.version';