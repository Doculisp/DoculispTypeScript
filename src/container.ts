import { IContainer, IDependencyContainer, IDependencyManager, IDictionary, IRegisterable, ITestableContainer, Valid } from "./types.containers";
import { globSync } from 'glob';

function findModules(container: IContainer) {
    let moduleNames = (
        globSync('./dist/**/*.js')
        .map(p => {
            return p.replace('dist', '.').replace('\\src', '').replace('.js', '').replaceAll('\\', '/');
        })
    );

    moduleNames.forEach(n => {
        import(n).then(m => {
            const keys = Object.keys(m);
            keys.
                filter(key => {
                    let value = m[key];
                    return !!value.name && !!value.builder;
                }).
                forEach(key => container.register(m[key]));
        });
    });
}

class Container implements ITestableContainer {
    _registry: IDictionary<IRegisterable> = {};
    _replacements: IDictionary<IRegisterable> = {};
    _cache: IDictionary<Valid<any>> = {};
    _repCache: IDictionary<Valid<any>> = {};
    _testable = false;

    constructor();
    constructor(registry: IDictionary<IRegisterable>);
    constructor(registry?: IDictionary<IRegisterable>) {
        this._testable = !!registry;
        if(!!registry) {
            const keys = Object.keys(registry);
            keys.forEach(key => this._registry[key] = registry[key]);
        }
        else {
            findModules(this);
        }
    }

    _hasKey(registry: IDictionary<IRegisterable>, key: string): boolean {
        return !!registry[key];
    }

    _register(registry: IDictionary<IRegisterable>, registerable: IRegisterable): ITestableContainer {
        let key = registerable.name;
        if (this._hasKey(registry, key)) {
            throw new Error(`Module named "${key}" already registered.`);
        }
        registry[key] = registerable;
        return this
    }

    register(registerable: IRegisterable): ITestableContainer {
        if (!registerable.name) {
            throw new Error('Must have a valid name to be registered.');
        }
        this._register(this._registry, registerable);
        return this;
    }
    
    registerValue(value: any, name?: string | undefined): ITestableContainer {
        if (!value.name && !name) {
            throw new Error('Most provide a name as a property or a parameter when registering a value as a module.');
        }

        const registerable: IRegisterable = {
            builder: () => { return value; },
            name: name ?? value.name,
            singleton: true
        };

        return this.register(registerable);
    }
    
    registerBuilder(fn: (...args: any[]) => any, dependencies: string[], name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if(!fn.name && !name) {
            throw new Error('Must provide a name on the function or as a parameter to register a builder.');
        }

        const registerable: IRegisterable = {
            builder: fn,
            name: name ?? fn.name,
            dependencies: dependencies ?? [],
            singleton: !!singleton,
        }
        
        this.register(registerable);
        return this;
    }

    _build<T>(moduleName: string, modules: string[]): Valid<T> {
        let moduleBuilder = this._replacements[moduleName];
        let replaced = true;
        if (!moduleBuilder) {
            moduleBuilder = this._registry[moduleName];
            replaced = false;
        }

        if (!moduleBuilder) {
            let t: any = undefined;
            try {
                t = require(moduleName);
            }
            catch {
            }

            if (!t) {
                throw new Error(`No module called "${moduleName}" registered`);
            }
            this.registerBuilder(() => require(moduleName), [], moduleName);
            moduleBuilder = this._registry[moduleName] as IRegisterable;
        }

        if(this._repCache[moduleName]) {
            return this._repCache[moduleName];
        }
        
        if(!replaced && this._cache[moduleName]) {
            return this._cache[moduleName];
        }

        let recursive = 
            moduleBuilder.dependencies?.filter(name => modules.includes(name)) ?? [];

        modules[modules.length] = moduleName;

        if (0 < recursive.length) {
            let moduleNames = modules.map(name => `"${name}"`).join(' => ');
            let recursiveModules = recursive.map(name => `"${name}"`).join(', ');
            throw new Error(`Circular dependencies between (${moduleNames} => [${recursiveModules}])`);
        }

        let dependencies: Valid<any>[] = 
            moduleBuilder.dependencies?.map((name) => this._build(name, modules)) ?? [];

        let result: Valid<T> = moduleBuilder.builder.apply(moduleBuilder.builder, dependencies);

        if(moduleBuilder.singleton) {
            if(replaced) {
                this._repCache[moduleName] = result;
            }
            else {
                this._cache[moduleName] = result;
            }
        }

        return result;
    }
    
    build(moduleName: string): Valid<any> {
        return this._build<any>(moduleName, []);
    }
    
    buildAs<T>(moduleName: string): Valid<T> {
        return this._build<T>(moduleName, []);
    }
    
    buildTestable(): ITestableContainer {
        return new Container(this._registry);
    }
    
    supportsReplace(): boolean {
        return this._testable;
    }

    _replace(registerable: IRegisterable, isPackage?: boolean) : ITestableContainer {
        let key = registerable.name;
        if(!isPackage && !this._hasKey(this._registry, key)) {
            throw new Error(`Cannot replace module "${key}" as it has not been registered.`);
        }
        if(!isPackage && this._hasKey(this._replacements, key)) {
            throw new Error (`Cannot replace module "${key}" as it has not been registered.`)
        }

        return this._register(this._replacements, registerable);
    }
    
    replace(registerable: IRegisterable): ITestableContainer {
        return this._replace(registerable);
    }
    
    replaceBuilder(fn: (...args: any[]) => any, dependencies: string[], name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if (!fn.name && !name) {
            throw new Error('Must provide a name either on the passed method or parameter to replace builder.');
        }

        let registerable: IRegisterable = {
            builder: fn,
            dependencies: dependencies ?? [],
            name: name ?? fn.name,
            singleton: !!singleton,
        };
        
        return this._replace(registerable);// 
    }
    
    replaceValue(value: any, name?: string | undefined): ITestableContainer {
        let key = name ?? value.name;
        if (!key) {
            throw new Error('Cannot replace value unless it has a name property or the name is passed as a parameter.');
        }

        return this.replaceBuilder(() => { return value; }, [], key, true);
    }
    
    replacePackageBuilder(fn: (...args: any[]) => any, name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if(!fn.name && !name) {
            throw new Error('Must provide a name either on the passed method or parameter to replace package builder.');
        }

        let registerable: IRegisterable = {
            builder: fn,
            name: name ?? fn.name,
            singleton: !!singleton,
        };
        
        return this._replace(registerable, true);
    }
    
    replacePackageValue(value: any, name?: string | undefined): ITestableContainer {
        if(!value.name && !name) {
            throw new Error('Cannot replace package with value unless it has a name property or the name is passed as a parameter.');
        }

        return this.replacePackageBuilder(() => value, name ?? value.name);
    }
    
    restore(moduleName: string): ITestableContainer {
        delete this._replacements[moduleName];
        delete this._repCache[moduleName];
        return this;
    }
    
    restoreAll(): ITestableContainer {
        const keys = Object.keys(this._replacements);
        let that = this;
        keys.forEach(key => that.restore(key));
        return this;
    }
    
    getModuleList(): string[] {
        const keys = Object.keys(this._registry);
        return Object.assign([], keys);
    }
}

const core = new Container();
const registry = core as IDependencyContainer;
const manager = core as IDependencyManager;
const container = core as IContainer;

export {
    registry,
    manager,
    container,
};