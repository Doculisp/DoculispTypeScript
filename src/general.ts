import { IRegisterable } from "./types.containers";
import { IFail, ILocation, ISuccess, IUtil } from "./types.general";

class Location implements ILocation {
    private readonly _line: number;
    private readonly _char: number;
    
    constructor(line: number, char: number) {
        this._line = line;
        this._char = char;
    }

    public get line(): number {
        return this._line;
    }

    public get char(): number {
        return this._char;
    }

    toString(): string {
        return `{ line: ${this._line}, char: ${this._char} }`;
    }

    asJson() {
        return { line: this._line, char: this._char };
    }
}

function buildGeneral(): IUtil {
    function ok<T>(successfulValue: T) : ISuccess<T> {
        return {
            value: successfulValue,
            success: true,
        };
    };

    function fail(message: string, documentPath: string) : IFail {
        return {
            message,
            documentPath,
            success: false,
        };
    };

    function location(line: number, char: number): ILocation {
        return new Location(line, char);
    }

    return {
        ok,
        fail,
        location,
    };
}

const astParser: IRegisterable = {
    builder: () => buildGeneral(),
    name: 'util',
    singleton: true,
    dependencies: []
};

export {
    astParser,
};