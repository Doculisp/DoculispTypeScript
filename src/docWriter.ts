import { AstBulletStyle, AstPart, IAst, IHeader, ILoad, ISectionWriter, ITableOfContents, ITitle, IWrite } from "./types.ast";
import { IRegisterable } from "./types.containers";
import { IDocWriter } from "./types.docWriter";
import { IUtil, Result } from "./types.general";

function buildWriter(util: IUtil): IDocWriter {
    function getLabel(style: AstBulletStyle, load: ILoad): string {
        switch (style) {
            case 'bulleted':
            case 'numbered':
            case 'unlabeled':
                return '';
            default:
                return load.sectionLabel;
        }
    }

    function getNumberedBullet(): (label: string, text: string) => string {
        let cnt = 1;
        function bullet(label: string, text: string): string {
            const lbl = 0 < label.length ? `${label}: ` : '';
            const result = `${cnt}. ${lbl}${text}`;
            cnt++;
            return result;
        }

        return bullet;
    }

    function getBullet(): (label: string, text: string) => string {
        function bullet(label: string, text: string): string {
            const lbl = 0 < label.length ? `${label}: ` : '';
            return `* ${lbl}${text}`;
        }

        return bullet;
    }

    function getNoBullet(): (label: string, text: string) => string {
        function bullet(label: string, text: string): string {
            const lbl = 0 < label.length ? `${label}: ` : '';
            return `${lbl}${text}`;
        }
    }

    function getBulletBuilder(style: AstBulletStyle): (label: string, text: string) => string {
        switch (style) {
            case 'bulleted':
            case 'bulleted-labeled':
                return getBullet();
            case 'numbered':
            case 'numbered-labeled':
                return getNumberedBullet();
            default:
                return getNoBullet();
        }
    }

    function compileToc(astToc: ITableOfContents, loads: ILoad): string {
        if(astToc.bulletStyle === 'no-table') {
            return '';
        }
    }

    function compileContentLocation(loads: ILoad[]): string {
        const parts: string[] = [];

        loads.forEach(load => {
            if(load.document) {
                const section = compileSection(load.document);
                parts.push(section);
            }
        });

        return `${parts.join('\n')}`;
    }

    function compileHeader(astHeader: IHeader): string {
        const padding = '#'.repeat(astHeader.depthCount);
        return `${padding} ${astHeader.text} ${padding}`;
    }

    function compileTitle(astTitle: ITitle): string {
        return astTitle.label;
    }

    function compileWrite(astWrite: IWrite): string {
        return astWrite.value;
    }

    function compilePart(astPart: AstPart) : string {
        astPart.
    }

    function compileSection(astSection: ISectionWriter): string {
        const parts: string[] = [];

        
    }

    function compile(astResult: Result<IAst>): Result<string> {
        if(!astResult.success) {
            return astResult;
        }

        const ast = astResult.value;

        if(ast.section.type === 'ast-empty') {
            return util.ok('');
        }

        throw new Error("Not yet Implemented");
    }

    function write(path: string, astResult: Result<IAst>): Result<string> {
        throw new Error('Not yet Implemented');
    }

    return {
        compile,
        write,
    };
}

const documentWriter: IRegisterable = {
    builder: (util: IUtil) => buildWriter(util),
    name: 'docWriter',
    singleton: true,
    dependencies: ['util', 'fileHandler']
};

export {
    documentWriter,
};