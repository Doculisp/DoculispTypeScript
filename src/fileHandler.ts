import { IRegisterable } from "./types.containers";
import { IFileHandler } from "./types.fileHandler";

import { IUtil, Result } from "./types.general";

function buildLoader(util: IUtil, fs: any): IFileHandler {
    function load(path: string): Result<string> {
        try {
            const value: string = fs.readFileSync(path, {encoding: 'utf8'});
            return util.ok(value);
        } catch (error) {
            return util.fail(`${error}`, path);
        }
    }

    return {
        load
    };
}

const loader: IRegisterable = {
    builder: (util: IUtil, fs: any) => buildLoader(util, fs),
    name: 'fileHandler',
    dependencies: ['util', 'fs'],
    singleton: true
};

export {
    loader,
};