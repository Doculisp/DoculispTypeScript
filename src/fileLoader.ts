import { IRegisterable } from "./types.containers";
import { IFileLoader } from "./types.fileLoader";

import { IUtil, Result } from "./types.general";

function buildLoader(util: IUtil, fs: any): IFileLoader {
    function load(path: string): Result<string> {
        try {
            const value: string = fs.readFileSync(path, {encoding: 'utf8'});
            return util.ok(value);
        } catch (error) {
            return util.fail(`${error}`, path);
        }
    }

    return {
        load
    };
}

const loader: IRegisterable = {
    builder: (util: IUtil, fs: any) => buildLoader(util, fs),
    name: 'fileLoader',
    dependencies: ['util', 'fs'],
    singleton: true
};

export {
    loader,
};