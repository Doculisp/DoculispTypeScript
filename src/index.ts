#! /usr/bin/env node
import { container } from './container';
import { Command } from 'commander';
import { Result } from './types/types.general';
import { IController } from './types/types.controller';
import { IVersion } from './types/types.version';
import figlet from 'figlet';

const program = new Command();
const controller = container.buildAs<IController>('controller');
const versionGetter = container.buildAs<IVersion>('version');

const versionMaybe = versionGetter.getVersion();


let version = '?.?.?'

if(!versionMaybe.success) {
    console.log('doculisp cannot determine its version')
} else {
    version = versionMaybe.value;
}

const helptext = '\n\n' + figlet.textSync('doculisp', {
    font: 'Cybermedium'
}) + `\n                     Version: ${version}`;

program.addHelpText('beforeAll', helptext);
program.addHelpText('afterAll', '\n\n');

program
    .name('doculisp')
    .description('A compiler for markdown')
    .version(version)
    .requiredOption('-s, --source <source_path>', 'the source file to compile')
    .requiredOption('-d, --output <output_path>', 'the output document path for the compiled markdown')
    .option('-t, --test', 'runs the compiler without generating the output file.')
    .parse(process.argv);

function reportResult(result: Result<string | false>) {
    if (result.success) {
        const message = !!result.value ? `Ok! ${result.value} successfully written` : 'Ok!';
        console.log(message);
    }
    else {
        console.log(JSON.stringify(result, null, 4));
    }
}

const options = program.opts();

const sourcePath: string = options['source'];

if (options['test']) {
    const result = controller.test(sourcePath);
    reportResult(result);
}
else {
    const outputPath: string = options['output'];
    const result = controller.compile(sourcePath, outputPath);
    reportResult(result);
}
