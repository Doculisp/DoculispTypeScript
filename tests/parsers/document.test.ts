import { container } from "../../src/container";
import { ITestableContainer } from "../../src/types.containers";
import { DocumentParser } from "../../src/types.document";
import { verifyAsJson } from "approvals/lib/Providers/Jest/JestApprovals";
import { configure } from "approvals/lib/config";
import { JestReporter } from "approvals/lib/Providers/Jest/JestReporter";

describe('document', () => {
    let environment: ITestableContainer = undefined as any;
    let parse: DocumentParser = undefined as any;

    beforeAll(() => {
        configure({
            reporters: [new JestReporter(), 'vscode'],
        });
    });

    beforeEach(() => {
        environment = container.buildTestable();
        parse = environment.buildAs<DocumentParser>('documentParse');
    });

    describe('parsing markup', () => {
        describe('text', () => {
            test('should successfully parse an empty string', () => {
                const result = parse('', 'C:/my_document.md')
        
                verifyAsJson(result);
            });
        
            test('should parse a simple text of "hello"', () => {
                const result = parse('hello', 'C:/my_document.md')
        
                verifyAsJson(result);
            });
    
            test('should parse text of "blow fish"', () => {
                const result = parse('blow fish', 'C:/my_document.md')
                verifyAsJson(result);
            });
    
            test('should parse text of " blow fish"', () => {
                const result = parse(' blow fish', 'C:/my_document.md')
                verifyAsJson(result);
            });
    
            test('should parse text of " blow fish "', () => {
                const result = parse(' blow fish ', 'C:/my_document.md')
                verifyAsJson(result);
            });
    
            test('should parse text of "   \\r\\n blow fish"', () => {
                const result = parse('   \r\n blow fish', 'C:/my_document.md');
                verifyAsJson(result);
            });
        });

        describe('html comments', () => {
            test('should not parse html comments', () => {
                const md = `<!-- This is a comment -->hello bro`.trim();
    
                const result = parse(md, 'C:/readme.md')
    
                verifyAsJson(result);
            });
    
            test('should not parse html but preserve new line counts comments', () => {
                const md = `<!--
    This is a comment
    -->
                     \t\thello bro
                     `;
    
                const result = parse(md, 'C:/readme.md')
    
                verifyAsJson(result);
            });
    
            test('should not parse html comments in the middle of text.', () => {
                const md = `hello
    <!-- need stuff here -->
    
    world`;
    
                const result = parse(md, 'C:/comments/helloWorld.md');
    
                verifyAsJson(result);
            });
    
            test('should parse html comments inside an inline code block', () => {
                const md = '`<!-- an example html comment -->`';
    
                const result = parse(md, 'C:/html/inline.md');
    
                verifyAsJson(result);
            });
    
            test('should parse html comments inside a multiline code block', () => {
                const md = `An example of an markdown document with html comments:
    \`\`\`markdown
    # A document
    
    <!-- this need
    a summary
    -->
    
    ## Sub section title
    \`\`\`
    `;
                const result = parse(md, 'C:/markdown/multiline.md');
    
                verifyAsJson(result);
            });
    
            test('should fail to parse if html comment is not closed', () => {
                let md = `<!--
    Hello
    World
    Boom
    `;
                const result = parse(md, 'C:/examples/bad.md');
    
                verifyAsJson(result);
            });
    
            test('should fail if inline code block does not close', () => {
                let md = '`let b = 7';
    
                const result = parse(md, 'C:/bad/noCloseInline.md');
    
                verifyAsJson(result);
            });
    
            test('should fail to parse an inline code block with a line break', () => {
                let md = `\`let a = 0;
    let b = a;
    \``;
    
                const result = parse(md, 'C:/examples/badInline.md');
    
                verifyAsJson(result);
            });
    
            test('should fail to parse a multiline code block that does not close', () => {
                let md = '```\nlet a = "hello;\nlet b = "world"\nconsole.log(a + " " + b);\n';
    
                const result = parse(md, 'C:/bad/examples/multiline.md');
    
                verifyAsJson(result);
            });
        });
    
        describe('Doculisp', () => {
            test('should parse a doculisp block at top of file', () => {
                const md = '<!-- (dl (# header)) -->';
    
                const result = parse(md, '_main.md');
    
                verifyAsJson(result);
            });
            
            test('should parse a doculisp block in the middle of file', () => {
            const md = '# Title\r\nsome text about title\r\n<!--\r\nSome lisp: (dl (# two)) -->\r\nMickey Mouse Hotline.';
    
                const result = parse(md, '_main.md');
    
                verifyAsJson(result);
            });
        });
    });

    describe('parsing .dlisp files', () => {
        test('should handle a correctly formatted file', () => {
            let dlisp = `
(section-meta
    (title Doculisp)
    (external
        (Section ./structure.md)
        (Section ./doculisp.md)
        (Section ./section-meta.md)
        (Section ./content.md)
        (Section ./headings.md)
        (Section ./comment.md)
        (Section ./keywords.md)
    )
)

(content (toc numbered-labeled))
`;

            let result = parse(dlisp, 'C:/main.dlisp');

            verifyAsJson(result);
        });
    });
});