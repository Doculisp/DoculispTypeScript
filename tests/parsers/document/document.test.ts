import { container } from "../../../src/container";
import { configure } from "approvals/lib/config";
import { ITestableContainer } from "../../../src/types.containers";
import { DocumentMap, DocumentParser } from "../../../src/types.document";
import { getVerifier } from "../../tools";
import { Options } from "approvals/lib/Core/Options";
import fs from 'fs';
import { IProjectLocation, Result } from "../../../src/types.general";

describe('document', () => {
    let environment: ITestableContainer = undefined as any;
    let parse: DocumentParser = undefined as any;
    let verifyAsJson: (data: any, options?: Options) => void;

    beforeAll(() => {
        verifyAsJson = getVerifier(configure);
    });

    beforeEach(() => {
        environment = container.buildTestable();
        parse = environment.buildAs<DocumentParser>('documentParse');
    });

    it('should not allow a document with a zero depth.', () => {
        const result = parse('hello', { documentPath: 'C:/my_document.md', documentDepth: 0, documentIndex: 6 });
        
        verifyAsJson(result);
    });

    it('should not allow a document with a negative depth.', () => {
        const result = parse('hello', { documentPath: 'C:/my_document.md', documentDepth: -1, documentIndex: 6 });
        
        verifyAsJson(result);
    });

    it('should not allow a document with a zero index.', () => {
        const result = parse('', { documentPath: 'C:/my_document.md', documentDepth: 4, documentIndex: 0 });
        
        verifyAsJson(result);
    });

    it('should not allow a document with a negative index.', () => {
        const result = parse('', { documentPath: 'C:/my_document.md', documentDepth: 4, documentIndex: -1 });
        
        verifyAsJson(result);
    });

    describe('parsing markup', () => {
        describe('text', () => {
            it('should successfully parse an empty string', () => {
                const result = parse('', { documentPath: 'C:/my_document.md', documentDepth: 4, documentIndex: 8 });
        
                verifyAsJson(result);
            });
        
            it('should parse a simple text of "hello"', () => {
                const result = parse('hello', { documentPath: 'C:/my_document.md', documentDepth: 3, documentIndex: 6 });
        
                verifyAsJson(result);
            });
    
            it('should parse text of "blow fish"', () => {
                const result = parse('blow fish', { documentPath: 'C:/my_document.md', documentDepth: 7, documentIndex: 2});
                verifyAsJson(result);
            });
    
            it('should parse text of " blow fish"', () => {
                const result = parse(' blow fish', { documentPath: 'C:/my_document.md', documentDepth: 4, documentIndex: 6 });
                verifyAsJson(result);
            });
    
            it('should parse text of " blow fish "', () => {
                const result = parse(' blow fish ', { documentPath: 'C:/my_document.md', documentDepth: 7, documentIndex: 1 });
                verifyAsJson(result);
            });
    
            it('should parse text of "   \\r\\n blow fish"', () => {
                const result = parse('   \r\n blow fish', { documentPath: 'C:/my_document.md', documentDepth: 1, documentIndex: 8 });
                verifyAsJson(result);
            });
        });

        describe('html comments', () => {
            it('should not parse html comments', () => {
                const md = `<!-- This is a comment -->hello bro`.trim();
    
                const result = parse(md, { documentPath: 'C:/readme.md', documentDepth: 5, documentIndex: 2 })
    
                verifyAsJson(result);
            });
    
            it('should not parse html but preserve new line counts comments', () => {
                const md = `<!--
    This is a comment
    -->
                     \t\thello bro
                     `;
    
                const result = parse(md, { documentPath: 'C:/readme.md', documentDepth: 8, documentIndex: 3 })
    
                verifyAsJson(result);
            });
    
            it('should not parse html comments in the middle of text.', () => {
                const md = `hello
    <!-- need stuff here -->
    
    world`;
    
                const result = parse(md, { documentPath: 'C:/comments/helloWorld.md', documentDepth: 1, documentIndex: 2 });
    
                verifyAsJson(result);
            });
    
            it('should parse html comments inside an inline code block', () => {
                const md = '`<!-- an example html comment -->`';
    
                const result = parse(md, { documentPath: 'C:/html/inline.md', documentDepth: 5, documentIndex: 3 });
    
                verifyAsJson(result);
            });
    
            it('should parse html comments inside a multiline code block', () => {
                const md = `An example of an markdown document with html comments:
    \`\`\`markdown
    # A document
    
    <!-- this need
    a summary
    -->
    
    ## Sub section title
    \`\`\`
    `;
                const result = parse(md, { documentPath: 'C:/markdown/multiline.md', documentDepth: 4, documentIndex: 3 });
    
                verifyAsJson(result);
            });
    
            it('should fail to parse if html comment is not closed', () => {
                let md = `<!--
    Hello
    World
    Boom
    `;
                const result = parse(md, { documentPath: 'C:/examples/bad.md', documentDepth: 5, documentIndex: 4 });
    
                verifyAsJson(result);
            });
    
            it('should fail if inline code block does not close', () => {
                let md = '`let b = 7';
    
                const result = parse(md, { documentPath: 'C:/bad/noCloseInline.md', documentDepth: 8, documentIndex: 4 });
    
                verifyAsJson(result);
            });
    
            it('should fail to parse an inline code block with a line break', () => {
                let md = `\`let a = 0;
    let b = a;
    \``;
    
                const result = parse(md, { documentPath: 'C:/examples/badInline.md', documentDepth: 6, documentIndex: 8 });
    
                verifyAsJson(result);
            });
    
            it('should fail to parse a multiline code block that does not close', () => {
                let md = '```\nlet a = "hello;\nlet b = "world"\nconsole.log(a + " " + b);\n';
    
                const result = parse(md, { documentPath: 'C:/bad/examples/multiline.md', documentDepth: 2, documentIndex: 7 });
    
                verifyAsJson(result);
            });
        });
    
        describe('Doculisp', () => {
            it('should parse a doculisp block at top of file', () => {
                const md = '<!-- (dl (# header)) -->';
    
                const result = parse(md, { documentPath: '_main.md', documentDepth: 3, documentIndex: 7 });
    
                verifyAsJson(result);
            });
            
            it('should parse a doculisp block in the middle of file', () => {
            const md = '# Title\r\nsome text about title\r\n<!--\r\nSome lisp: (dl (# two)) -->\r\nMickey Mouse Hotline.';
    
                const result = parse(md, { documentPath: '_main.md', documentDepth: 7, documentIndex: 5 });
    
                verifyAsJson(result);
            });

            it('should parse lisp outside an html tag as text', () => {
                const md = '(# Heading) Hello Doculisp';

                const result = parse(md, { documentPath: 'documentExample.md', documentDepth: 3, documentIndex: 4 });

                verifyAsJson(result);
            });

            it('should parse Doculisplisp outside an html tag as text', () => {
                const md = '(dl (# Heading)) Hello Doculisp';

                const result = parse(md, { documentPath: 'documentExample2.md', documentDepth: 8, documentIndex: 8 });

                verifyAsJson(result);
            });
        });
    });

    describe('parsing .dlisp files', () => {
        it('should handle a correctly formatted file', () => {
            let dlisp = `
(section-meta
    (title Doculisp)
    (external
        (Section ./structure.md)
        (Section ./doculisp.md)
        (Section ./section-meta.md)
        (Section ./content.md)
        (Section ./headings.md)
        (Section ./comment.md)
        (Section ./keywords.md)
    )
)

(content (toc numbered-labeled))
`;

            let result = parse(dlisp, { documentPath: 'C:/main.dlisp', documentDepth: 7, documentIndex: 1 });

            verifyAsJson(result);
        });

        it('should fail to parse a file that contains a dl atom', () => {
            let dlisp = `(dl
(section-meta
    (title Doculisp)
)

(content (toc numbered-labeled)))
`;

            let result = parse(dlisp, { documentPath: 'C:/bad/extraDl.dlisp', documentDepth: 5, documentIndex: 6 });

            verifyAsJson(result);
        });

        it('should handle a file with parentheses that do not close', () => {
            let dlisp = `
(section-meta
    (title Doculisp)
    (external
        (Section ./structure.md)
        (Section ./doculisp.md
        (Section ./section-meta.md)
    )
)
`;

            let result = parse(dlisp, { documentPath: 'C:/main.dlisp', documentDepth: 6, documentIndex: 1 });

            verifyAsJson(result);
        });

        it('should handle a file with to many parenthesis', () => {
            let dlisp = `(content (toc numbered-labeled)) )`;

            let result = parse(dlisp, { documentPath: 'C:/main.dlisp', documentDepth: 5, documentIndex: 8 });

            verifyAsJson(result);
        });
    });

    describe('parsing real files', () => {
        it('should parse the content of structure.md first inline error', () =>{
            const result = parse(`<!--
(dl
    (section-meta
        (title Basic Structure)
    )
)
-->

The basic structure of Doculisp is all code is contained within blocks. A block is constructed within an HTML comment region. It starts with an open parentheses \`(\`
`, { documentPath: './structure.md', documentDepth: 2, documentIndex: 1 });
        
        verifyAsJson(result);
        });

        it('should parse the content of structure.md second inline error', () =>{
            const result = parse(`<!--
(dl
    (section-meta
        (title Basic Structure)
    )
)
-->

The basic structure of Doculisp is all code is contained within blocks. A block is constructed within an HTML comment region. It starts with an open parentheses \`(\` followed by a sting of non-whitespace characters. This is called an atom. It then has 1 of three possibilities. It can have a parameter, a new block, or nothing. All blocks must close with a close parentheses \`)\`.

Even the Doculisp main block follows this.

Example

\`\`\`markdown
<!--
(dl
    (section-meta
        (title Basic Structure)
    )
)
-->
\`\`\`

The first block is the \`dl\` block. In it \`dl\` is the atom. It contains the \`section-meta\` sub-block.  That block has the atom \`section-meta\` followed by a further sub block. The last sub block is the \`title\` sub block. In it \`title\` is the atom and \`Basic Structure\`
`, { documentPath: './structure.md', documentDepth: 2, documentIndex: 1 });
        
        verifyAsJson(result);
        });

        function getContent(fileName: string, depth: number, index: number): Result<DocumentMap> {
            const path: string = `./documentation/${fileName}`;
            const location: IProjectLocation = {
                documentDepth: depth,
                documentIndex: index,
                documentPath: path,
            };

            const content = parse(fs.readFileSync(path, { encoding: 'utf8' }), location);
            return content;
        }

        it('should parse the content of structure.md from the file system', () => {
            const result = getContent('structure.md', 2, 1);
            verifyAsJson(result);
        });

        it('should parse the content of doculisp.md from the file system', () => {
            const result = getContent('doculisp.md', 2, 2);
            verifyAsJson(result);
        });

        it('should parse the content of section-meta.md from the file system', () => {
            const result = getContent('section-meta.md', 2, 3);
            verifyAsJson(result);
        });

        it('should parse the content of content.md from the file system', () => {
            const result = getContent('content.md', 2, 4);
            verifyAsJson(result);
        });

        it('should parse the content of headings.md from the file system', () => {
            const result = getContent('headings.md', 2, 5);
            verifyAsJson(result);
        });

        it('should parse the content of comment.md from the file system', () => {
            const result = getContent('comment.md', 2, 6);
            verifyAsJson(result);
        });

        it('should parse the content of keywords.md from the file system', () => {
            const result = getContent('keywords.md', 2, 7);
            verifyAsJson(result);
        });

        it('should parse the content of _main.dlisp from the file system', () => {
            const result = getContent('_main.dlisp', 1, 1);
            verifyAsJson(result);
        });
    });
});