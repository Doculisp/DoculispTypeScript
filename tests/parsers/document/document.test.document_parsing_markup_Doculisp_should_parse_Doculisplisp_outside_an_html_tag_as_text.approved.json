{
  "success": true,
  "value": {
    "parts": [
      {
        "location": {
          "char": 1,
          "documentDepth": 8,
          "documentIndex": 8,
          "documentPath": "documentExample2.md",
          "line": 1
        },
        "text": "(dl (# Heading)) Hello Doculisp",
        "type": "text"
      }
    ],
    "projectLocation": {
      "documentDepth": 8,
      "documentIndex": 8,
      "documentPath": "documentExample2.md"
    }
  }
}
